twilio-auth-sample/src/main/java/com/twilio/authsample/main/TokensFragment.java

```diff
@@ -46,24 +51,57 @@ private void initViews(View rootView) {
 
         LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
         recyclerView.setLayoutManager(layoutManager);
+        tokensAdapter = new TokensAdapter(new ArrayList<AuthenticatorToken>(), this);
+        recyclerView.setAdapter(tokensAdapter);
+    }
+
+    @Override
+    public void onTokenClicked(AuthenticatorToken app) {
+        Intent intent = new Intent(getActivity(), TokenDetailsActivity.class);
+        intent.putExtra(TokenDetailsActivity.EXTRA_APP_ID, app.getAppId());
+        intent.putExtra(Intent.EXTRA_TITLE, app.getName());
+
+        getActivity().startActivity(intent);
     }
 
     @Override
     public void onResume() {
         super.onResume();
+        twilioAuthenticator.addObserver(this);
 
-        List<AuthenticatorToken> apps = twilioAuthenticator.getApps();
+    }
 
-        tokensAdapter = new TokensAdapter(apps, this);
-        recyclerView.setAdapter(tokensAdapter);
+    @Override
+    public void onPause() {
+        twilioAuthenticator.removeObserver(this);
+        super.onPause();
     }
 
     @Override
-    public void onTokenClicked(AuthenticatorToken app) {
-        Intent intent = new Intent(getActivity(), TokenDetailsActivity.class);
-        intent.putExtra(TokenDetailsActivity.EXTRA_APP_ID, app.getAppId());
-        intent.putExtra(Intent.EXTRA_TITLE, app.getName());
+    public void onTOTPsUpdated(@NonNull TOTPs totps) {
+        Set<String> appIds = totps.keySet();
+        List<AuthenticatorToken> authenticatorTokens = new ArrayList<>();
+        for (String appId : appIds) {
+            authenticatorTokens.add(totps.getTOTP(appId));
+        }
+        tokensAdapter.setApps(authenticatorTokens);
+    }
 
-        getActivity().startActivity(intent);
+    @Override
+    public void onAppAdded(@NonNull AuthenticatorToken app) {
+        tokensAdapter.addApp(app);
+        tokensAdapter.notifyDataSetChanged();
```

Comment:
Shouldn't this be done directly by the adapter? When would you like to add an app and not notify the change?

Reply:
totally agree

Reply:
done